hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/Listing.java

```diff
@@ -390,8 +392,18 @@ public S3AFileStatus next() throws IOException {
         status = statusBatchIterator.next();
         // We remove from provided list the file status listed by S3 so that
         // this does not return duplicate items.
-        if (providedStatus.remove(status)) {
-          LOG.debug("Removed the status from provided file status {}", status);
+
+        // The provided status is returned as it is assumed to have the better
+        // metadata (i.e. the eTag and versionId from S3Guard)
+        Optional<S3AFileStatus> provided =
```

Comment:
why not just say 
```
status2 = providedStatus.remove(status);
if (status2 != null) {
 return status;
}

That way: no need to scan the list looking for a entry when the remove operation will be doing it anyway?
```

Reply:
OK, I see now. Set.remove() doesn't return the old version, only that there was an old version